app/Actions/Photo/Strategies/AddBaseStrategy.php

```diff
@@ -0,0 +1,153 @@
+<?php
+
+namespace App\Actions\Photo\Strategies;
+
+use App\Exceptions\JsonError;
+use App\Facades\AccessControl;
+use App\Models\Logs;
+use App\Models\Photo;
+
+abstract class AddBaseStrategy
+{
+	protected AddStrategyParameters $parameters;
+	protected Photo $photo;
+
+	protected function __construct(AddStrategyParameters $parameters, Photo $photo)
+	{
+		$this->parameters = $parameters;
+		$this->photo = $photo;
+	}
+
+	abstract public function do(): Photo;
+
+	/**
+	 * Hydrates meta-info of the media file from the
+	 * {@link AddStrategyParameters::$info} attribute of the associated
+	 * {@link AddStrategyParameters} object into the associated {@link Photo}
+	 * object.
+	 *
+	 * Meta information is conditionally copied if and only if the target
+	 * attribute of the {@link Photo} object is null or empty and the
+	 * meta-info is not.
+	 * This way this method is usable by {@link AddStandaloneStrategy} and
+	 * {@link AddDuplicateStrategy}.
+	 * For a freshly created {@link Photo} object (with empty attributes)
+	 * all available meta-data is hydrated, but for an already existing
+	 * {@link Photo} object existing attributes are not overwritten.
+	 */
+	protected function hydrateMetadata()
+	{
+		if (empty($this->photo->title) && !empty($this->parameters->info['title'])) {
```

Comment:
Can't this be refactored slightly ?
The code is pretty much the same all the time, so with a for loop over `['title', 'description' etc...]` it could be slightly more compact? :thinking:

Reply:
I totally agree. It can. However, this is just cut & paste code from before. I did not come up with this code by myself.

I actually would like to drop the associative array (which is returned by the metadata extractor) completely and replace that associative array with a proper class. This would yield three benefits: 1) proper attributes, 2) "type safety" (as far as possible in PHP) 3) we could use a proper class method as a conversion routine which would replace that code above.

However, again I wanted to postpone that to a future PR because dropping the array and replacing it by a proper object would also touch the Metadata extractor and this PR is already huge enough.